lib/OrdinaryDiffEqSDIRK/src/generic_sdirk_perform_step.jl

```diff
@@ -0,0 +1,478 @@
+    
+    # handle additional stages if present
+    z₂ = nothing
+    if s >= 2
+        if hasfield(typeof(cache), :z₂)
+            z₂ = getfield(cache, :z₂)
+            @.. broadcast=false z₂ = zero(eltype(u))
+            nlsolver.z = z₂
+            @.. broadcast=false nlsolver.tmp = uprev + A[2,1] * z₁
+            nlsolver.c = typeof(nlsolver.c)(c[2])
+            nlsolver.γ = typeof(nlsolver.γ)(A[2,2])
+            
+            isnewton(nlsolver) && set_new_W!(nlsolver, false)
+            z₂ .= nlsolve!(nlsolver, integrator, cache, repeat_step)
+            nlsolvefail(nlsolver) && return
+        end
+    end
+    
+    # computes final solution
+    if s == 1
+        @.. broadcast=false u = uprev + b[1] * z₁
+    elseif s == 2 && z₂ !== nothing
+        @.. broadcast=false u = uprev + b[1] * z₁ + b[2] * z₂
+    else
+        error("Unsupported number of stages: $s")
+    end
+    
+    step_limiter!(u, integrator, p, t + dt)
+    
+    # error estimation for adaptive methods
+    if integrator.opts.adaptive && b_embed !== nothing
+        if s == 1
+            @.. broadcast=false tmp = b_embed[1] * z₁
+        elseif s == 2 && z₂ !== nothing
+            @.. broadcast=false tmp = b_embed[1] * z₁ + b_embed[2] * z₂
+        end
+        
+        has_smooth_est = hasfield(typeof(alg), :smooth_est)
+        if has_smooth_est && alg.smooth_est && isnewton(nlsolver)
+            est = atmp
+            linres = dolinsolve(integrator, nlsolver.cache.linsolve; b = _vec(tmp),
+                linu = _vec(est))
+            integrator.stats.nsolve += 1
+        else
+            est = tmp
+        end
+        
+        calculate_residuals!(atmp, est, uprev, u, integrator.opts.abstol,
+            integrator.opts.reltol, integrator.opts.internalnorm, t)
+        integrator.EEst = integrator.opts.internalnorm(atmp, t)
+    end
+    
+    # set fsallast based on last stage
+    if s == 1
+        @.. broadcast=false integrator.fsallast = z₁ / dt
+    elseif s == 2 && z₂ !== nothing
+        @.. broadcast=false integrator.fsallast = z₂ / dt
+    end
+end
+
+@muladd function generic_additive_sdirk_perform_step!(integrator, cache::SDIRKConstantCache, repeat_step=false)
+    @unpack t, dt, uprev, u, f, p = integrator
+    @unpack tab, nlsolver = cache
+    alg = unwrap_alg(integrator, true)
+    
+    if integrator.f isa SplitFunction
+        f_impl = integrator.f.f1
+        f_expl = integrator.f.f2
+    else
+        f_impl = integrator.f
+    end
+    
+    markfirststage!(nlsolver)
+    
+    s = size(tab.A, 1)
+    γ = tab.γ
+    A = tab.A
+    b = tab.b
+    c = tab.c
+    b_embed = tab.b_embed
+    
+    A_explicit = tab.A_explicit
+    b_explicit = tab.b_explicit
+    
+    if integrator.f isa SplitFunction && !repeat_step && !integrator.last_stepfail
+        z₁ = dt * f_impl(uprev, p, t)
+    else
+        z₁ = dt * integrator.fsalfirst
+    end
+    
+    if s == 4  # 4-stage methods like KenCarp3, Kvaerno3
+        z₂ = z₁
```

Comment:
This is not generally a good prediction? It should be a linear combination from tableau coefficients?